### PR TITLE
NFS host and guest support

### PR DESCRIPTION
**Pull request by djdv**
Indirectly requested from https://github.com/ipfs/roadmap/issues/83
Creating a draft before the bot locks the thread.

This is still a WIP but basic functionality seems to work for both hosting and attaching.
However, initial testing shows it is not reliable.
Looking at the logs, there seems to be some kind of continuity problem in the NFS server or client where 2 lookup requests back to back will be processed in the wrong order. (Likely some kind of race condition.)
```
2023/11/21 13:21:51 [TRACE] request: RPC #4290061661 (nfs.Lookup)
2023/11/21 13:21:51 [TRACE] request: RPC #4290061662 (nfs.Lookup)
err on: Qmc2fSA6chjZMbXRi7JGQZ7tXEmDbtG2XF9FXRLCtGLBQV [] xid did not mat ch, expected: ffb5255e, received: ffb5255d
err on: Qmc2fSA6chjZMbXRi7JGQZ7tXEmDbtG2XF9FXRLCtGLBQV [] xid did not mat ch, expected: ffb5255d, received: ffb5255e
```
The error stems from [this section of the client](https://github.com/willscott/go-nfs-client/blob/271fa9065b3377f1124f3ce74f957d8a2dd3ab0c/nfs/rpc/client.go#L106-L108).
and I noticed the server is handling sends on the socket within a goroutine (see references to [this struct member](https://github.com/willscott/go-nfs/blob/a91b8b3bf0fd76882bb2f2a2f3c1256405c03f15/conn.go#L41)).
The issue could lie within the NFS libraries or my own implementations, so I'll have to investigate that when I get the chance.

---

Example server/host invocation:
`fs mount nfs pinfs /ip4/192.168.1.40/tcp/2049`
I.e. `fs mount nfs $guestAPI $nfsServerMaddr`
(Assuming an IPFS node is running on this machine with default settings. Otherwise pass `-pinfs-api $ipfsAPIMaddr` before the final multiaddr argument.)

Example guest/client invocation:
`fs mount fuse nfs -nfs-server /ip4/192.168.1.40/tcp/2049 I:`
I.e. `fs mount $host nfs $nfsServerMaddr $mountPoint`

---

CC: @willscott
We spoke about utilizing this library for this purpose a very long time ago but I never got around to it until now.
Letting you know about its existence.
I'm planning on looking into the above error myself sometime, but you're more than welcome to hack around too. ;^]

### CODE REVIEW

**Comment by willscott** [comment]
Looking briefly at the client you linked I think that's right - the client is only coded to support 1 call occurring / outstanding at a time.

**Comment by djdv** [comment]
Had some time to look at this.
Ended up just making the client lock between operations to prevent any overlapping requests. (And generally fixing up and finishing other parts.)

Still a few things left to do but the server and client seem to work. At least well enough for a smoke test.
https://www.youtube.com/watch?v=19FkIxTzavY

**Comment by willscott** [comment]
try updating to the latest v0.0.2 of go-nfs and you may see the stale handle issue you ran into in the video go away.

This is very cool!

**Comment by djdv** [comment]
>try updating to the latest v0.0.2 of go-nfs and you may see the stale handle issue you ran into in the video go away.

Did a quick test and it seems like it did fix that. :^]
Some very minor changes to get it to build again, but afterwards the Go server and Linux client seemed to work alright.

However, https://github.com/willscott/go-nfs/pull/121 broke some stuff  on my end.
So I'm going to look into patching this branch to comply with the changes.

>This is very cool!

Thanks!

---

Extra technical details on the breakage:

On the server side, within [mountpoint.go:Mount(...)](https://github.com/djdv/go-filesystem-utils/blob/4402e351d321294b45d505107ee93e7f5e89997f/internal/filesystem/nfs/mountpoint.go#L53), I'm basically taking an extended `fs.FS` and wrapping that in a [`billy.Basic` implementation](https://github.com/djdv/go-filesystem-utils/blob/4402e351d321294b45d505107ee93e7f5e89997f/internal/filesystem/nfs/server.go#L18C4-L18C4), which gets passed to `polyfill`, before being wrapped further with NFS specifics (`NewNullAuthHandler`+`NewCachingHandler`), and finally starts listening for connections.

The `fs.FS` implementations I've written, don't (yet) implement `Lstat`, so when I make `target.Getattr` calls from `go-nfs-client`[^1]), they're returning `NFS3ERR_NOTDIR` (specifically the logs say: `[ERROR] call to 0x7ff6e6f6e0a0 failed: Not a directory: lstat ".": unsupported operation` which is coming from the [`fs.FS` `Lstat` extension function I have](https://github.com/djdv/go-filesystem-utils/blob/4402e351d321294b45d505107ee93e7f5e89997f/internal/filesystem/filesystem.go#L160).[^2])

To remedy this I can just implement `Lstat` methods on all the `fs.FS` implementations I have.
And also add a check for this up front in `mountpoint.go:Mount`, returning a clear error stating we require an `Lstat` method.
This is related to the last part in https://github.com/djdv/go-filesystem-utils/issues/41 expressing the need for something similar to billy's `Capabilities` concept so that different abstraction layers can better interoperate.

[^1]: I do this [in the `fs.FS`' `Open` method](https://github.com/djdv/go-filesystem-utils/blob/4402e351d321294b45d505107ee93e7f5e89997f/internal/filesystem/nfs/client.go#L271)
[^2]: [polyfill's `Lstat` method](https://github.com/go-git/go-billy/blob/9f60ea941235eed8357c6adbe01addeeefe76398/helper/polyfill/polyfill.go#L76) also behaves this way.

**Comment by djdv** [comment]
Added symbolic link support to the IPFS `fs.FS` systems.
Specifically `Lstat`, and `Readlink`; some systems can conditionally do `Symlink` (`mklink`) as well.
`Open` should now implicitly try to resolve links, as long as they are relative to the `fs.FS` itself.
Symlink support is not well tested yet, but these changes restore NFS support on  `go-nfs` v0.2, which includes the fix for Linux's client.
We could have got away with just doing `Lstat`, but proper symlink support had to happen eventually anyway.

Other link related changes were made in other parts of the code too (filesystem interface pkg + FUSE host).
The NFS client implementation imposes similar link restrictions (as above). It will do a best effort inspection on link targets and try to resolve ones which are relative.
There's a constructor option that allows setting a path delimiter token, so if you know your links will be delimited by a token other than `/` (such as `somewhere\relative`) you can construct the FS [with an option](https://github.com/djdv/go-filesystem-utils/blob/d864f3ac62db438637de33b6aec5ccb6a8b97c10/internal/filesystem/nfs/client.go#L153-L168) that tells it to [replace them in the target string](https://github.com/djdv/go-filesystem-utils/blob/d864f3ac62db438637de33b6aec5ccb6a8b97c10/internal/filesystem/nfs/client.go#L263-L265). So that they get split properly instead of treated as a filename character (as opposed to a path separator)

When I get a chance I'm going to clean up some remaining TODOS, and do some testing with IPFS link nodes.
After that should just need to review it and squash some commits.

**Comment by willscott** [comment]
happy to do a review pass when you're ready for it

**Comment by djdv** [comment]
This is probably good enough for a first pass.
Will, I can't flag you as a reviewer, so I added you as a collaborator first.
If you accept that I should be able to put you on this one. :^]

Some context on various packages:

`/filesystem` contains extensions to `fs.FS`, adding any functionality not present in the standard, with extension functions to match the interface extensions. All of our "guest" systems intend to implement parts of these interfaces.

Its subpkg `/filesystem/errors` essentially extends `fs.PathError`, adding a `Kind` type. Guests can define a general error "kind" which hosts APIs can inspect and translate into their native error values.
E.g. an [IPFS guest might return an fserror](https://github.com/djdv/go-filesystem-utils/blob/604993b6b346d74d83d3f36e94c1efa446489884/internal/filesystem/ipfs/ipfs.go#L185) which can be picked up by a FUSE host, that [can translate it into](https://github.com/djdv/go-filesystem-utils/blob/604993b6b346d74d83d3f36e94c1efa446489884/internal/filesystem/cgofuse/translate.go#L178) the specific `errno` value its API requires.

 `/nfs` contains file system host & guest support utilizing the extensions in `/filesystem`.
`/ipfs` are guest implementations that can be used by other host implementations. They've been adapted to utilize symlinks in this patchset.

All the other changes could be ignored, but you're welcome to look at them too if you want.
Extra stuff below:

***

The general idea for `/commands/mount.go` is that it parses `argv` and transforms values into structures that hold parameters for the host and guest constructors. Those get marshaled and sent over the wire to a daemon process, which itself unmarshals the data, and constructs the guest `fs.FS` and passes it to the host instances `Mount` method.
There are reasons for doing things that way, but the implementation is too complicated as-is. So I'm planning on refactoring it later to reduce some duplication and other funny business.

There will still be a need to keep "mountpoint" data serializable, but I think it can be done better than it currently is.
1 reason is that we need to [send our CLI args to the daemon process](https://github.com/djdv/go-filesystem-utils/blob/master/doc/development.md#client---server---os-interactions), and 2 is that we need to be able to store the mountpoint on disk so that we can restore it later, like Unix's `fstab`.
There's other interesting things you can do when the mount table is a living service in itself, such as changing instance parameters live, with or without needing to remount / migrate handles. As well as integrations with external tools, for example a context menu being able to know it's inside an IPFS mount point, and offer more functions like "copy CID", "copy gatewaylink" like an [old shell-extension of mine](https://user-images.githubusercontent.com/157609/47924095-4c96c500-debb-11e8-9e2e-9d7342f610fa.gif)
Some of this already works: 
[Desktop 2023.02.03 - 13.25.43.13.mp4](https://www.youtube.com/watch?v=ojclkBtIAEU&list=PLrZxGPqQuwHAevUMSBHTyRVnB6h5WOu5x&index=38&t=168s), [Desktop 2022.08.23 - 18.14.54.08.mp4](https://www.youtube.com/watch?v=r-_U8RcOWZE&list=PLrZxGPqQuwHAevUMSBHTyRVnB6h5WOu5x&index=39))
but isn't well documented since it's likely going to change.

File: internal/commands/mountpoint_nfs.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+		}
+		return "", errors.New("NFS host address was not present in the mountpoint data")
+	}
+}
+
+func makeNFSGuestCommand[
+	HC mountCmdHost[HT, HM],
+	HM marshaller,
+	HT any,
+](host filesystem.Host,
```
Comment (willscott) [inline internal/commands/mountpoint_nfs.go:96]: do you need the linebreak here after the return?

File: internal/commands/mountpoint_nfs.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+	"github.com/djdv/go-filesystem-utils/internal/command"
+	"github.com/djdv/go-filesystem-utils/internal/filesystem"
+	p9fs "github.com/djdv/go-filesystem-utils/internal/filesystem/9p"
+	"github.com/djdv/go-filesystem-utils/internal/filesystem/nfs"
+	"github.com/djdv/go-filesystem-utils/internal/generic"
+	"github.com/multiformats/go-multiaddr"
+)
+
+type (
```
Comment (willscott) [inline internal/commands/mountpoint_nfs.go:20]: the generics and the combination of host and guest mount in this file make the logic a bit hard to follow - i wonder if some of the generics could be normalized to have a bit more repetition but be more readable.

File: internal/filesystem/ipfs/ipns.go [hunk 3]
```diff
@@ -63,6 +64,7 @@ func NewIPNS(core coreiface.CoreAPI, ipfs fs.FS, options ...IPNSOption) (*IPNS,
 					readAll | executeAll,
 			},
 			nodeTimeout: 1 * time.Minute,
+			linkLimit:   40, // Arbitrary.
```
Comment (willscott) [inline internal/filesystem/ipfs/ipns.go:67]: raise to constant somewhere?

File: internal/filesystem/ipfs/shared.go [hunk 4]
```diff
@@ -392,3 +471,67 @@ func fsTypeName(mode fs.FileMode) string {
...
+			err  = generic.ConstError("link target must be relative")
+			kind = fserrors.InvalidItem
+		)
+		pair := fmt.Sprintf(
+			`%s -> %s`,
+			name, target,
+		)
+		return "", fserrors.New(op, pair, err, kind)
+	}
+	if target = path.Join("/"+name, target); target == "/" {
```
Comment (willscott) [inline internal/filesystem/ipfs/shared.go:533]: does this work on windows?

File: internal/filesystem/nfs/client.go [hunk 5]
```diff
@@ -0,0 +1,611 @@
...
+	if err := settings.fillDefaults(); err != nil {
+		return nil, err
+	}
+	naddr, err := manet.ToNetAddr(maddr)
+	if err != nil {
+		return nil, err
+	}
+	const (
+		network    = "tcp"
+		privledged = false
```
Comment (willscott) [inline internal/filesystem/nfs/client.go]:
```suggestion
		privileged = false
```

File: internal/filesystem/nfs/client.go [hunk 6]
```diff
@@ -0,0 +1,611 @@
...
+	"github.com/willscott/go-nfs-client/nfs"
+	"github.com/willscott/go-nfs-client/nfs/rpc"
+)
+
+type (
+	goFS struct {
+		target        *nfs.Target
+		linkSeparator string
+		linkLimit     uint
+		// NOTE [2024.01.08]: The NFS server library is able to handle multiple requests concurrently
```
Comment (willscott) [inline internal/filesystem/nfs/client.go:27]: the lock is unfortunate - there may be a better nfs client out there that doesn't have the concurrency limitation.

File: internal/filesystem/nfs/mountpoint.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+	// (Primarily to suppress `ENOENT` errors in the console.)
+	const nfslibEnvKey = "LOG_LEVEL"
+	if _, set := os.LookupEnv(nfslibEnvKey); !set {
+		nfs.Log.SetLevel(nfs.PanicLevel)
+	}
+	go func() { errsCh <- nfs.Serve(goListener, cachedHandler) }()
+	return closerFn, nil
+}
+
+func (*Guest) GuestID() filesystem.ID { return GuestID }
```
Comment (willscott) [inline internal/filesystem/nfs/mountpoint.go:94]: this means you're only able to have one mount of this at a time because it'll always have the same global filesystem ID?

File: internal/commands/mountpoint_nfs.go (hunk 2 above)
Comment (djdv) [reply internal/commands/mountpoint_nfs.go:20]: Agreed. 
There are some benefits to having these generic command constructors, but even as the author I get tripped up on it. Especially with some of the bad constraint names, and huge signatures.
I've been considering alternate ways to deal with it but have been struggling to come up with something that is both good and compiler legal. 
Thinking about trying to do a large refactor after this PR that splits a lot of this up, but I'm worried about creating some kind of dependency cycle somewhere. We'll see how it goes.

File: internal/filesystem/ipfs/ipns.go (hunk 3 above)
Comment (djdv) [reply internal/filesystem/ipfs/ipns.go:67]: For sure. This is part of a larger problem as well, detailed below but can be ignored.

---

The way I handle mountpoint data (the parameters required to instantiate a file system) is currently weird.
And I was thinking about better ways to handle them.

The `ipfs` pkg has constructors for the IPFS `fs.FS` systems, but in `/command/mount` we create a struct containing values for that constructor.
(The mount command is just a client that sends that struct + other data to the `fs` daemon process, which translates that portion back into parameters for the `fs.FS` constructor which is calls in its own process space).

I'm putting thought into how to better deal with this, but I'm not sure yet.
One thing I was considering was just changing the constructors to accepting a settings struct, and exposing that. This way the constructor parameter is itself an exposed serializable type/format.
The other was going to be to keep them as-is, but expose all the default values in exported consts `DefaultLinkLimit = 40`, so that external packages could come up with their own serializable format if they need to, irrelevant of how the Go API is.
Or some other nonsense.

The main reason the constructor options need to be serializable, is so that we can implement something similar to Unix's `fstab`. We can just dump the current mountpoints to disk, and load them again at startup.
(This currently works, but there's no command to automate it. I want to fix this up first before doing that.)

File: internal/filesystem/ipfs/shared.go (hunk 4 above)
Comment (djdv) [reply internal/filesystem/ipfs/shared.go:533]: In this context it's not supposed to.
Go's `fs.FS` paths are non-rooted `/` separated paths.
And this call is only used in an `fs.FS` context.
When our system resolves links, it wants that path format.
The `ReadLink` method is an exception, which returns the raw symlink string no matter what it is.
But we don't use that method to traverse links internally.

That said, I do think there's a bug here.
I'm rooting the path at `/`, and joining it, in case there's any dots that climb all the way up to the `path` root (`/`), and if so converting it to the `fs.FS` root (`.`).
But I'm never taking that `/` back out, which creates an invalid `fs.FS` path.
I.e. `/go/fs/standard` should actually be `go/fs/standard`.

File: internal/commands/mountpoint_nfs.go (hunk 1 above)
Comment (djdv) [reply internal/commands/mountpoint_nfs.go:96]: I think this might be `gofumpt` rules being enabled in my LSP settings.
If I remove it, it just inserts it again.
I like most of the decisions of `gofumpt`, but it usually does weird stuff with function signatures...
I'll leave this alone for now and hope the entire function gets blown away in a future refactor.

File: internal/filesystem/nfs/client.go (hunk 5 above)
Comment (djdv) [reply internal/filesystem/nfs/client.go]: (I'll probably end up squashing this before merge and just attribute you as Co-Author on the whole client commit.)

File: internal/filesystem/nfs/client.go (hunk 6 above)
Comment (djdv) [reply internal/filesystem/nfs/client.go:27]: For sure.
I'd be open to suggestions on alternatives.
But it might also be worth filing an issue about it and hoping it gets resolved eventually.
Can't commit to looking into it myself right now, but it could happen eventually. Of course anyone else could pop in for it too.

File: internal/filesystem/nfs/client.go (hunk 5 above)
Comment (djdv) [reply internal/filesystem/nfs/client.go]: lol, I didn't update any references to this. I'll fix it.

File: internal/filesystem/nfs/mountpoint.go (hunk 7 above)
Comment (djdv) [reply internal/filesystem/nfs/mountpoint.go:94]: Nah. It's just the name of the file system.
That interface method is probably poorly name. It used to be `FSID` iirc. Maybe it should be `.Type` or something else?

I forget all the places it's used, but I think the mount table uses it. The mount commands might use it, but I think not for "reasons".

Check this out, if we start the daemon with an exposed socket, it acts as a 9P server.
We can then instantiate a bunch of mount points, then as a 9P client, see what the actual mount table looks like.
Which shows those names being used in both the paths and mountpoint tags.
```
Windows:
fs daemon -api-server /ip4/192.168.1.40/tcp/564

fs mount nfs keyfs -api-server /ip4/192.168.1.40/tcp/564 /ip4/192.168.1.40/tcp/2050
fs mount nfs pinfs -api-server /ip4/192.168.1.40/tcp/564 /ip4/192.168.1.40/tcp/2051

fs mount nfs pinfs -api-server /ip4/192.168.1.40/tcp/564 /ip4/192.168.1.40/tcp/2049
fs.exe mount fuse nfs -api-server /ip4/192.168.1.40/tcp/564 -nfs-server /ip4/192.168.1.40/tcp/2049 I:

Linux:
mount -t 9p 192.168.1.40 /mnt/9 -o trans=tcp,port=564
tree /mnt/9
```
Screenshot 2024-02-06 105930

File: internal/filesystem/ipfs/shared.go (hunk 4 above)
Comment (djdv) [reply internal/filesystem/ipfs/shared.go:533]: Resolved the issue where I was returning a `/` rooted path (that `fs.FS` does not want)
https://github.com/djdv/go-filesystem-utils/compare/01526aa45558472b148a6ea0309a84ec0d84a0a7..841152d326564906a4a0db301ac10af143bff37d

File: internal/filesystem/nfs/client.go (hunk 6 above)
Comment (djdv) [reply internal/filesystem/nfs/client.go:27]: FWIW it looks like the issue tracker for [your fork](https://github.com/willscott/go-nfs-client) is disabled.
If that's intentional, no problem.

Comment (willscott) [reply internal/filesystem/nfs/client.go:27]: issues now enabled

**Comment by djdv** [comment]
I'm planning to look over these changes again when I get a chance, and afterwards will probably merge this in.
If people use this and find bugs with it of course they can be patched later.

Will, thanks for the help on this. I appreciate you going beyond just the NFS parts too.
I think we both agree that some of the parts of the codebase are complex and need some attention.
I'm not sure yet how much it can be improved, but I'm going to at least investigate it and prototype some stuff.
But that's going to be done in a separate PR later.

I'm in no rush to get this merged, so if you're still looking at the code, or plan on running some interactive tests against it, just let me know and I'll hold off on pressing the button.

**Comment by willscott** [comment]
> I'm in no rush to get this merged, so if you're still looking at the code, or plan on running some interactive tests against it, just let me know and I'll hold off on pressing the button.

I wasn't planning to look at this again in the short term. Thanks for proving it out though, I'm excited it exists!

**Comment by djdv** [comment]
Sorry for the silence on this.
Instead of reviewing it and merging it ASAP, I ended up trying to refactor a lot of the code related to mount.
(I've also been tied up with unrelated troubles.)

I did end up doing a rather large refactor of the code base, but I'm not actually sure if it's an improvement.
As such, I'm going to finish up that refactor, try to get this merged, then try to get the refactor merged after that.

Afterwards, I'm planning on recording a lengthy video orientation, with each section explaining each package of the code base, along with rationale for why it was done the way it is.
The purpose of this is twofold, 1) it should help people to understand how the program works in case they want to modify it, and 2) act as an open invitation for suggestions.
If someone watches it and thinks there's a better way to implement some part of the codebase, I'd be happy to hear it since this repo is rather complicated and I'm having a hard time reducing that complexity.

Personal note: I have to be gone for a few days but I'll act on this when I get back.

---

Edit (2024.03.19):
Still working on a large refactor but I have to disappear for a few days again.
Have a temporary commit up in this branch: [j/mount-refactor](https://github.com/djdv/go-filesystem-utils/tree/j/mount-refactor), but it's probably not worth looking at yet.
Didn't have time to split the commits up yet so it's one giant blob for now. I'll remedy that later.

I did port the FUSE host and the IPFS guests over to a newer style on that branch. And most of the command line flag stuff was redone as well.
That branch is built on top of this one, but I still need to port the NFS host and guest support to the newer interfaces there.

Intended to have that finished this week, but got interrupted with something IRL again. Hopefully next week.
When I posted previously, the current iteration of the refactor was not looking much better. But I think the one I landed on is probably somewhat better. Hard to gauge it - since I wrote it, it makes sense to me right now but we'll see in time if the code holds up in terms of clarity.
